[streams] Add tests for aborting pipeTo()

ricea opened the pull request:
Add tests for aborting ReadableStream.prototype.pipeTo using an
AbortSignal.

Tests correspond to PR whatwg/streams#744.

ricea committed:
[streams] Add tests for aborting pipeTo()

Add tests for aborting ReadableStream.prototype.pipeTo using an
AbortSignal.

Tests correspond to PR whatwg/streams#744.

ricea committed:
Fix lint errors

ricea committed:
Update wrapper abort.html

domenic commented on streams/piping/abort.js:
```diff
@@ -0,0 +1,239 @@
+      }
+    });
+    const ws = new WritableStream();
+    return promise_rejects(t, new TypeError(), rs.pipeTo(ws, { signal: invalidSignal }), 'pipeTo should reject');
```
Should we test that this never calls any underlying source methods?

domenic commented on streams/piping/abort.js:
```diff
@@ -0,0 +1,239 @@
+      .then(() => {
+        assert_equals(rs.events.length, 2, 'cancel should have been called');
+        assert_equals(rs.events[0], 'cancel', 'first event should be cancel');
+        assert_equals(rs.events[1].name, 'AbortError', 'the argument to cancel should be an AbortError');
```
Maybe also test that its .constructor.name is DOMException.

ricea replied:
Done.

domenic reviewed (approved):
Overall LGTM, some minor suggestions. A bit amused that almost all of the tests use an already-aborted signal, but I guess that's where the edge cases lie.

OK, here's a suggestion: wait for the pipe to complete successfully, then abort the signal, and make sure that doesn't somehow error/cancel the streams? I dunno, maybe that's too silly.

ricea committed:
Verify that nothing is done when signal is invalid

Also check the constructor name of the AbortError is DOMException.

ricea commented:
I will write tests for abort-after-completion. That was an important oversight.

ricea committed:
Test abort behaviour after shutdown condition

Once a shutdown condition happens, aborting the signal should no longer
do anything. Verify this works.

ricea committed:
Test that pending writes are completed on abort

ricea committed:
Strengthen assertions that abort does nothing after pipe finishes

Use preventCancel, preventClose and preventAbort to keep the readable or
writable open after the pipeTo finishes to verify that the abort doesn't
close them.
